<h1 id="december-12-2011">December 12, 2011</h1>
<ul>
<li>Added data from the TCGA colorectal cancer project</li>
<li>Added support for RNA-Seq based expression data (colorectal cancer)</li>
<li>Added support for log2 copy-number data (available on the Plots tab)</li>
<li>Updated GBM copy-number, expression, and methylation data</li>
</ul>
<h1 id="november-14-2011">November 14, 2011</h1>
<ul>
<li>New and improved mutation details, with sorting and filtering capabilities.</li>
<li>In collaboration with Bilkent University, we have added a new <em>network</em> tab to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
</ul>
<p><img alt="GBM Network" src="images/previews/ova_network.png" /></p>
<h1 id="september-3-2011">September 3, 2011</h1>
<ul>
<li>You can now query across different cancer studies (feature available directly from the home page).</li>
<li>Our <a href="cgds_r.jsp">MATLAB CGDS Cancer Genomics Toolbox</a> is now available.  The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<p><img alt="Cross Cancer Query" src="images/previews/cross_cancer.png" /></p>
<h1 id="march-2-2011">March 2, 2011</h1>
<p>New plotting features and other improvements:</p>
<ul>
<li>Correlation plots that show the relationship between different data types for individual genes.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Updated <a href="cgds_r.jsp">R Package</a> with support for correlation plots and general improvements for retrieving and accessing data in R data frames.</li>
<li>The <a href="web_api.jsp">Web Interface</a> now supports basic clinical data, e.g. survival data.</li>
<li><a href="networks.jsp">Networks</a> for pathway analysis are now available for download.</li>
</ul>
<p><img alt="Survival Analysis" src="images/previews/ova_survival_60_percent.png" /></p>
<h1 id="december-15-2010">December 15, 2010</h1>
<p>Several new features, including:</p>
<ul>
<li>Redesigned and streamlined user interface, based on user feedback and usability testing.</li>
<li>Advanced support for gene-specific alterations.  For example, users
can now view mutations within TP53, and ignore copy number alterations, or
only view amplifications of EGFR, and ignore deletions.</li>
<li>Improved performance.</li>
<li><a href="faq.jsp">Frequently Asked Questions</a> document released.</li>
<li>Updated <a href="video.jsp">Video Tutorial</a>.</li>
</ul>
<h1 id="november-4-2010">November 4, 2010</h1>
<ul>
<li>Enhanced <a href="faq.jsp#what-are-oncoprints">Oncoprints</a>, enabling users to quickly visualize genomic alterations across many cases.  Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.</li>
<li>Official release of our <a href="web_api.jsp">Web Interface</a>, enabling programmatic access to all data.</li>
<li>Official release of our <a href="cgds_r.jsp">R Package</a>, enabling programmatic access to all data from the R platform for statistical computing.</li>
</ul>
<p><img alt="OncoPrints" src="images/previews/gbm_oncoprint.png" /></p>