 [TOC]

# What is the cBio Cancer Genomics Portal?

The portal stores genomic data from large scale, integrated cancer genomic data sets. It allows explorative data analysis (e.g.: Is my gene of interest altered in a specific cancer type? How frequently is EGFR amplified in glioblastoma? Do mutations of BRCA1 and BRCA2 in ovarian cancer co-occur?) and provides simple download of small data slices (user-defined gene and sample sets, no need to download entire data sets).

# How do I get started?

We have recently posted [two mini-tutorials](tutorial.jsp) to get you up and running.

# What data types are in the portal?

The portal currently stores DNA copy-number data (putative, discrete values per gene, e.g. "homozygously deleted" or "amplified"), mRNA and microRNA expression data, non-synonymous mutations, protein-level and phosphoprotein level (RPPA) data, and limited clinical data related to survival.

# How is the cBio Cancer Genomics Portal different from the TCGA Data Portal?

The cBio portal is an exploratory analysis tool for exploring large-scale cancer genomic data sets.  You can quickly view genomic alterations across a set of patients, across a set of cancer types, perform survival analysis and perform network analysis.  By contrast, the [TCGA Data Portal](http://tcga-data.nci.nih.gov/tcga/tcgaHome2.jsp) aims to be the definitive place for full-download and access to all data generated by TCGA.  If you want to explore a pathway of interest in one or more cancer types, the cBio portal is probably where you want to start.  However, if you want to download raw mRNA expression files or full segmented copy number files, the TCGA Data Portal is probably where you want to start.

# Why do some cancer studies have mutation data and others do not?

We store mutation data for published cancer studies, such as the TCGA Glioblastoma and TCGA Ovarian.
We do not, however store mutation data for provisional cancer data sets generated by TCGA.  This is because provisional studies contain candidate somatic mutations only, and as per NCI guidelines, we are not allowed to redistribute candidate somatic mutation data until such mutations have been validated.  As each cancer study is published and finalized by the TCGA, we will import the corresponding mutation data.

# Does the portal contain cancer study X?

Check out the [Data Sets Page](data_sets.jsp) for the complete set of cancer studies currently stored in the portal.  If you do not see your specific cancer study of interest, please contact us directly, and we will let you know if it's in the queue.

# What kind of clinical data is stored in the portal?

The portal currently stores overall and disease-free survival data, when available.

# Does the portal contain methylation data for all tumor types?

No, methylation data was not generated for the prostate cancer and sarcoma studies. We are working on adding methylation data for the GBM study, and it is available for the ovarian cancer study.

# Is methylation data available for all genes?

We only include methylation data from probes that show anti-correlation between the methylation signal and the gene's expression. For genes with multiple probes, we include the one with the strongest anti-correlation.

# Does the portal store raw or probe-level data?

No, the portal only contains gene-level data. Data for different isoforms of a given gene are merged. Raw and probe-level data for all date sets is available via NCBI GEO or through the TCGA Data Portal. See the cancer type description on the main query page for links to the raw data.

# What are OncoPrints?

OncoPrints are compact means of visualizing distinct genomic alterations, including somatic mutations, copy number alterations, and mRNA expression changes across a set of cases.  They are extremely useful for visualizing gene set and pathway alterations across a set of cases, and for visually identifying trends, such as trends in mutual exclusivity or co-occurence between gene pairs within a gene set. Individual genes are represented as rows, and individual cases or patients are represented as columns. ![Example OncoPrint](images/oncoprint_example.png)

# Can I change the order of genes in the OncoPrint?

The order of genes in the OncoPrint is determined by the order entered into the initial query field. Simply change the initial gene order, resubmit your query, and the change will be reflected in the OncoPrint.

# Does the portal work on all browsers and operating systems?

We support and test on the following web browsers:  Internet Explorer 7.0 and above, Firefox 3.0 and above, Safari and Google Chrome.  Note hover that SVG images, which are used for downloadable OncoPrints, are not supported by Internet Explorer versions older than 9. If you notice any other incompatibilities, please let us know.

# What are mRNA and microRNA Z-Scores?

For mRNA and microRNA expression data, we typically compute the relative expression of an individual gene and tumor to the gene's expression distribution in a reference population. That reference population is either all tumors that are diploid for the gene in question, or, when available, normal adjacent tissue. The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). This measure is useful to determine whether a gene is up- or down-regulated relative to the normal samples or all other tumor samples.

# What do "-2", "-1", "0", "1", and "2" mean in the copy-number data?

These levels are derived from the copy-number analysis algorithms GISTIC or RAE, and indicate the copy-number level per gene. "-2" is a deep loss, possibly a homozygous deletion, "-1" is a single-copy loss (heterozygous deletion), "0" is diploid, "1" indicates a low-level gain, and "2" is a high-level amplification. Note that these calls are putative.

# What is GISTIC?  What is RAE?

Copy number data sets within the portal are generated by [GISTIC](http://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](http://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837)
algorithms.  Both algorithms attempt to identify significantly altered regions of amplification or deletion across sets of patients. Both algorithms also generate putative gene/patient copy number specific calls, which are then input into the portal.  

# How do I analyze my own data sets in the cBio Portal?

Currently, we do not support uploading of user data sets into the cBio Portal.  However, the cBio Portal is open source software and you may want to consider deploying your own local instance of the cBio portal.  This may require some assistance from a system administrator.  To ease installation, we have also created a [Amazon Machine Image (AMI) for rapid deployment to Amazon Web Services](http://code.google.com/p/cbio-cancer-genomics-portal/wiki/AmazonEc2AMI).  A complete set of administration documents are available on our [wiki](http://code.google.com/p/cbio-cancer-genomics-portal/w/list)

# Is the cBio Portal open source?

Yes.  The cBio Portal is available under a [GNU Lesser General Public License](http://www.gnu.org/licenses/lgpl.html).  You can download our code and sample data sets at [Google code](http://code.google.com/p/cbio-cancer-genomics-portal/).  If you are interested in contributing, please contact us directly.

# How do I cite the portal?

We are finalizing a portal paper right now.  In the meantime, you can cite us by URL:  The cBio Cancer Genomics Portal (http://www.cbioportal.org/).

# How do I get updates on new portal developments and new data sets?

The best option is to follow us on [Twitter](https://twitter.com/cbioportal).  We tweet each time we add a new feature or a new data set.